File: code/language-and-framework-comparisons/language-comparison/markup-renderer/markup_renderer.pl
use strict;
use warnings;

sub render_markup {
    my ($text) = @_;

    my @markup_tags = (
        [qr/\[title\](.*?)\[\/title\]/, "\e[33;1m$1\e[0m"],     # Yellow bold text
        [qr/\[emphasis\](.*?)\[\/emphasis\]/, "\e[31;1m$1\e[0m"] # Red bold text
    );

    foreach my $tag (@markup_tags) {
        my ($regex, $replacement) = @{$tag};
        $text =~ s/$regex/$replacement/g;
    }

    return $text;
}

sub main {
    my $input_filename = "input.txt";
    open(my $file, '<', $input_filename) or die "File '$input_filename' not found.";

    local $/; # Slurp mode
    my $content = <$file>;
    close($file);

    my $rendered_content = render_markup($content);
    print $rendered_content;
}

main();

File: code/language-and-framework-comparisons/language-comparison/markup-renderer/markup_renderer.ml
open Str

let render_markup text =
  let markup_tags = [
    ("\\[title\\]\\(.*?\\)\\[/title\\]", "\027[33;1m\\1\027[0m");   (* Yellow bold text *)
    ("\\[emphasis\\]\\(.*?\\)\\[/emphasis\\]", "\027[31;1m\\1\027[0m") (* Red bold text *)
  ] in
  List.fold_left (fun txt (tag, replacement) ->
    global_replace (regexp tag) replacement txt
  ) text markup_tags

let () =
  let input_filename = "input.txt" in
  try
    let content = Stdlib.(input_line (open_in input_filename)) in
    let rendered_content = render_markup content in
    print_endline rendered_content
  with Sys_error _ ->
    print_endline ("File '" ^ input_filename ^ "' not found.")
